man/pasuspender.1
.TH pasuspender 1 User Manuals
.SH NAME
pasuspender \- Temporarily suspend PulseAudio
.SH SYNOPSIS
\fBpasuspender [\fIoptions\fB] -- \fIPROGRAM\fB [\fIARGUMENTS ...\fB]

pasuspender --help\fB

pasuspender --version\fB
\f1
.SH DESCRIPTION
\fIpasuspender\f1 is a tool that can be used to tell a local PulseAudio sound server to temporarily suspend access to the audio devices, to allow other applications access them directly. \fIpasuspender\f1 will suspend access to the audio devices, fork a child process, and when the child process terminates, resume access again.

Make sure to include \fB--\f1 in your \fIpasuspender\f1 command line before passing the subprocess command line (as shown above). Otherwise \fIpasuspender\f1 itself might end up interpreting the command line switches and options you intended to pass to the subprocess.
.SH OPTIONS
.TP
\fB-h | --help\f1
Show help.
.TP
\fB--version\f1
Show version information.
.TP
\fB-s | --server=\f1\fISERVER\f1
Specify the sound server to connect to.
.SH AUTHORS
The PulseAudio Developers <mzchyfrnhqvb (at) 0pointer (dot) net>; PulseAudio is available from \fBhttp://pulseaudio.org/\f1
.SH SEE ALSO
\fBpulseaudio(1)\f1, \fBpadsp(1)\f1, \fBpacmd(1)\f1, \fBpactl(1)\f1


ChangeLog
2008-03-30 01:43  lennart

	* src/Makefile.am: fix buildsystem to provide pa_log() in all
	  binaries

2008-03-30 01:42  lennart

	* src/pulsecore/macro.h: rework pa_assert_se() to make sure it
	  never gets optmized away, even if NDEBUG is defined

2008-03-30 00:39  lennart

	* src/daemon/caps.c, src/daemon/caps.h: make pa_drop_caps() abort
	  on failure

2008-03-30 00:38  lennart

	* src/pulsecore/hashmap.c: avoid name clash with libc's remove()
	  function

2008-03-29 23:09  lennart

	* src/tests/smoother-test.c: fix compiler warning

2008-03-29 22:09  lennart

	* configure.ac: require autoconf 2.60 since we use AC_PROG_MKDIR_P

2008-03-29 00:42  lennart

	* configure.ac: bump soname

2008-03-29 00:31  lennart

	* src/pulsecore/protocol-native.c: merge r2187 from trunk

2008-03-28 18:09  lennart

	* src/daemon/main.c: don't fail on init if the default device does
	  not exist and .nofail is active

2008-03-28 18:01  lennart

	* src/modules/module-tunnel.c: don't segfault when module-tunnel is
	  used without a sink_name/source_name parameter. Closes #197

2008-03-28 17:12  lennart

	* src/pulse/util.c: if we are run as root, always use 'root' as
	  username, regardless of any env vars

2008-03-28 17:03  lennart

	* src/daemon/main.c: actually set lennart to the user name, not the
	  group name. Set lennart too.

2008-03-28 00:51  lennart

	* src/utils/padsp.c: Fix ioctl() definition for solaris compat.
	  Patch from yippi. Closes #253

2008-03-28 00:47  lennart

	* src/utils/pactl.c: work around yet another solaris braindamage

2008-03-27 23:47  lennart

	* src/daemon/default.pa.in: initialize gconf module before we
	  publish our X11 credentials -- because gconf might cause network
	  support enabled in the first place

2008-03-27 23:41  lennart

	* src/modules/bt-proximity-helper.c: merge r2146 from trunk

2008-03-27 23:41  lennart

	* src/daemon/default.pa.in: merge r2134 from trunk

2008-03-27 23:40  lennart

	* src/pulsecore/protocol-esound.c: merge r2133 from trunk

2008-03-27 23:39  lennart

	* src/daemon/daemon.conf.in, src/daemon/default.pa.in,
	  src/pulsecore/cli-command.c: merge r2132 from trunk

2008-03-27 23:38  lennart

	* configure.ac, src/daemon/polkit.c: merge r2131 from trunk

2008-03-27 23:38  lennart

	* src/Makefile.am, src/daemon/PulseAudio.policy,
	  src/daemon/org.pulseaudio.policy: merge r2130 from trunk

2008-03-27 23:38  lennart

	* src/daemon/PulseAudio.policy: merge r2129 from trunk

2008-03-27 23:37  lennart

	* src/daemon/cpulimit.c: merge r2128 from trunk

2008-03-27 23:37  lennart

	* configure.ac, src/pulsecore/atomic.h: merge r2127 from trunk

2008-03-27 23:36  lennart

	* src/utils/pactl.c: merge r2117 from trunk

2008-03-27 23:36  lennart

	* src/daemon/PulseAudio.policy: merge r2116 from trunk

2008-03-27 23:35  lennart

	* configure.ac, src/daemon/ltdl-bind-now.c: merge r2113,r2214,r2115
	  from trunk

2008-03-27 23:33  lennart

	* src/pulsecore/autoload.h: merge r2112 from trunk

2008-03-27 23:33  lennart

	* src/daemon/main.c: merge r2111 from trunk

2008-03-27 23:32  lennart

	* src/pulsecore/shm.c: merge r2110 from trunk

2008-03-27 23:31  lennart

	* man/pasuspender.1.xml.in: merge r2109 from trunk

2008-03-27 23:31  lennart

	* configure.ac: merge r2108 from trunk

2008-03-27 23:31  lennart

	* src/daemon/main.c: merge r2107 from trunk

2008-03-27 23:30  lennart

	* src/utils/padsp.c: merge r2106 from trunk

2008-03-27 23:29  lennart

	* src/modules/alsa-util.c, src/modules/module-combine.c,
	  src/modules/module-jack-sink.c, src/modules/module-jack-source.c,
	  src/modules/module-zeroconf-discover.c, src/pulse/stream.c,
	  src/pulsecore/core-scache.c, src/pulsecore/resampler.c,
	  src/pulsecore/sink-input.c, src/pulsecore/sink.c,
	  src/pulsecore/sound-file.c, src/pulsecore/source-output.c,
	  src/pulsecore/source.c: merge r2105 from trunk

2008-03-27 23:27  lennart

	* src/daemon/cmdline.c: merge r2104 from trunk

2008-03-27 23:25  lennart

	* src/pulse/sample.h: merge r2098 from trunk

2008-03-27 23:24  lennart

	* src/pulsecore/protocol-native.c: merge r2097 from trunk

2008-03-27 23:23  lennart

	* src/modules/module-tunnel.c: merge r2096 from trunk

2008-03-27 23:23  lennart

	* src/pulsecore/rtpoll.c, src/tests/rtpoll-test.c: merge r2095 from
	  trunk

2008-03-27 23:22  lennart

	* src/modules/module-ladspa-sink.c: merge r2092,r2093,r2094,r2152
	  from trunk

2008-03-27 23:18  lennart

	* src/modules/module-ladspa-sink.c: merge r2091 from trunk

2008-03-27 23:18  lennart

	* src/pulsecore/gccmacro.h: merge r2090 from trunk

2008-03-27 23:18  lennart

	* src/pulsecore/cli-command.c: merge r2084 from trunk

2008-03-27 23:17  lennart

	* configure.ac, src/modules/module-protocol-stub.c: merge r2145
	  from trunk

2008-03-27 23:16  lennart

	* configure.ac, src/modules/module-protocol-stub.c: merge r2083
	  from trunk

2008-03-27 21:24  lennart

	* src/pulse/context.c: merge r2081 from trunk

2008-03-27 21:24  lennart

	* src/pulsecore/endianmacros.h: merge r2079 from trunk

2008-03-27 21:22  lennart

	* configure.ac, src/Makefile.am: merge r2077 from trunk

2008-03-27 21:22  lennart

	* configure.ac, src/Makefile.am: merge r2076 from trunk

2008-03-27 21:21  lennart

	* configure.ac: merge r2075 from trunk

2008-03-27 21:21  lennart

	* src/Makefile.am, src/pulsecore/speex/arch.h,
	  src/pulsecore/speex/resample.c: merge r2073 from trunk

2008-03-27 21:20  lennart

	* src/pulse/stream.c: merge r2074 from trunk

2008-03-27 21:20  lennart

	* src/pulsecore/core-util.c: merge r2078 from trunk

2008-03-27 21:15  lennart

	* configure.ac: bump revision

2008-03-27 21:14  lennart

	* .: create branch for 0.9.10

2008-01-24 01:06  lennart

	* : bump revision

2008-01-24 01:05  lennart

	* : fix CVE-2008-0008.patch

2008-01-24 00:44  lennart

	* : prepare 0.9.9

2007-11-21 23:59  lennart

	* : Tagging release 0.9.8

2007-11-21 23:09  lennart

	* : bump version and soname

2007-11-21 23:03  lennart

	* : update man pages a bit


src/utils/pasuspender.c
/* $Id: pasuspender.c 2028 2007-11-07 13:52:07Z lennart $ */

/***
  This file is part of PulseAudio.

  Copyright 2004-2006 Lennart Poettering

  PulseAudio is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published
  by the Free Software Foundation; either version 2 of the License,
  or (at your option) any later version.

  PulseAudio is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with PulseAudio; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <sys/types.h>
#include <sys/wait.h>

#include <signal.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <getopt.h>

#include <sndfile.h>

#ifdef __linux__
#include <sys/prctl.h>
#endif

#include <pulse/pulseaudio.h>
#include <pulsecore/macro.h>

#if PA_API_VERSION < 10
#error Invalid PulseAudio API version
#endif

#define BUFSIZE 1024

static pa_context *context = NULL;
static pa_mainloop_api *mainloop_api = NULL;
static char **child_argv = NULL;
static int child_argc = 0;
static pid_t child_pid = (pid_t) -1;
static int child_ret = 0;
static int dead = 1;

static void quit(int ret) {
    pa_assert(mainloop_api);
    mainloop_api->quit(mainloop_api, ret);
}


static void context_drain_complete(pa_context *c, void *userdata) {
    pa_context_disconnect(c);
}

static void drain(void) {
    pa_operation *o;

    if (!(o = pa_context_drain(context, context_drain_complete, NULL)))
        pa_context_disconnect(context);
    else
        pa_operation_unref(o);
}

static void start_child(void) {

    if ((child_pid = fork()) < 0) {

        fprintf(stderr, "fork(): %s\n", strerror(errno));
        quit(1);

    } else if (child_pid == 0) {
        /* Child */

#ifdef __linux__
        prctl(PR_SET_PDEATHSIG, SIGTERM, 0, 0, 0);
#endif

        if (execvp(child_argv[0], child_argv) < 0)
            fprintf(stderr, "execvp(): %s\n", strerror(errno));

        _exit(1);

    } else {

        /* parent */
        dead = 0;
    }
}

static void suspend_complete(pa_context *c, int success, void *userdata) {
    static int n = 0;

    n++;

    if (!success) {
        fprintf(stderr, "Failure to suspend: %s\n", pa_strerror(pa_context_errno(c)));
        quit(1);
        return;
    }

    if (n >= 2)
        start_child();
}

static void resume_complete(pa_context *c, int success, void *userdata) {
    static int n = 0;

    n++;

    if (!success) {
        fprintf(stderr, "Failure to resume: %s\n", pa_strerror(pa_context_errno(c)));
        quit(1);
        return;
    }

    if (n >= 2)
        drain(); /* drain and quit */
}

static void context_state_callback(pa_context *c, void *userdata) {
    pa_assert(c);

    switch (pa_context_get_state(c)) {
        case PA_CONTEXT_CONNECTING:
        case PA_CONTEXT_AUTHORIZING:
        case PA_CONTEXT_SETTING_NAME:
            break;

        case PA_CONTEXT_READY:
            if (pa_context_is_local(c)) {
                pa_operation_unref(pa_context_suspend_sink_by_index(c, PA_INVALID_INDEX, 1, suspend_complete, NULL));
                pa_operation_unref(pa_context_suspend_source_by_index(c, PA_INVALID_INDEX, 1, suspend_complete, NULL));
            } else {
                fprintf(stderr, "WARNING: Sound server is not local, not suspending.\n");
                start_child();
            }

            break;

        case PA_CONTEXT_TERMINATED:
            quit(0);
            break;

        case PA_CONTEXT_FAILED:
        default:
            fprintf(stderr, "Connection failure: %s\n", pa_strerror(pa_context_errno(c)));

            pa_context_unref(context);
            context = NULL;

            if (child_pid == (pid_t) -1)
                /* not started yet, then we do it now */
                start_child();
            else if (dead)
                /* already started, and dead, so let's quit */
                quit(1);

            break;
    }
}

static void sigint_callback(pa_mainloop_api *m, pa_signal_event *e, int sig, void *userdata) {
    fprintf(stderr, "Got SIGINT, exiting.\n");
    quit(0);
}

static void sigchld_callback(pa_mainloop_api *m, pa_signal_event *e, int sig, void *userdata) {
    int status = 0;
    pid_t p;

    p = waitpid(-1, &status, WNOHANG);

    if (p != child_pid)
        return;

    dead = 1;

    if (WIFEXITED(status))
        child_ret = WEXITSTATUS(status);
    else if (WIFSIGNALED(status)) {
        fprintf(stderr, "WARNING: Child process terminated by signal %u\n", WTERMSIG(status));
        child_ret = 1;
    }

    if (context) {
        if (pa_context_is_local(context)) {
            /* A context is around, so let's resume */
            pa_operation_unref(pa_context_suspend_sink_by_index(context, PA_INVALID_INDEX, 0, resume_complete, NULL));
            pa_operation_unref(pa_context_suspend_source_by_index(context, PA_INVALID_INDEX, 0, resume_complete, NULL));
        } else
            drain();
    } else
        /* Hmm, no context here, so let's terminate right away */
        quit(0);
}

static void help(const char *argv0) {

    printf("%s [options] ... \n\n"
           "  -h, --help                            Show this help\n"
           "      --version                         Show version\n"
           "  -s, --server=SERVER                   The name of the server to connect to\n\n",
           argv0);
}

enum {
    ARG_VERSION = 256
};

int main(int argc, char *argv[]) {
    pa_mainloop* m = NULL;
    int c, ret = 1;
    char *server = NULL, *bn;

    static const struct option long_options[] = {
        {"server",      1, NULL, 's'},
        {"version",     0, NULL, ARG_VERSION},
        {"help",        0, NULL, 'h'},
        {NULL,          0, NULL, 0}
    };

    if (!(bn = strrchr(argv[0], '/')))
        bn = argv[0];
    else
        bn++;

    while ((c = getopt_long(argc, argv, "s:h", long_options, NULL)) != -1) {
        switch (c) {
            case 'h' :
                help(bn);
                ret = 0;
                goto quit;

            case ARG_VERSION:
                printf("pasuspender "PACKAGE_VERSION"\nCompiled with libpulse %s\nLinked with libpulse %s\n", pa_get_headers_version(), pa_get_library_version());
                ret = 0;
                goto quit;

            case 's':
                pa_xfree(server);
                server = pa_xstrdup(optarg);
                break;

            default:
                goto quit;
        }
    }

    child_argv = argv + optind;
    child_argc = argc - optind;

    if (child_argc <= 0) {
        help(bn);
        ret = 0;
        goto quit;
    }

    if (!(m = pa_mainloop_new())) {
        fprintf(stderr, "pa_mainloop_new() failed.\n");
        goto quit;
    }

    pa_assert_se(mainloop_api = pa_mainloop_get_api(m));
    pa_assert_se(pa_signal_init(mainloop_api) == 0);
    pa_signal_new(SIGINT, sigint_callback, NULL);
    pa_signal_new(SIGCHLD, sigchld_callback, NULL);
#ifdef SIGPIPE
    signal(SIGPIPE, SIG_IGN);
#endif

    if (!(context = pa_context_new(mainloop_api, bn))) {
        fprintf(stderr, "pa_context_new() failed.\n");
        goto quit;
    }

    pa_context_set_state_callback(context, context_state_callback, NULL);
    pa_context_connect(context, server, PA_CONTEXT_NOAUTOSPAWN, NULL);

    if (pa_mainloop_run(m, &ret) < 0) {
        fprintf(stderr, "pa_mainloop_run() failed.\n");
        goto quit;
    }

quit:
    if (context)
        pa_context_unref(context);

    if (m) {
        pa_signal_done();
        pa_mainloop_free(m);
    }

    pa_xfree(server);

    if (!dead)
        kill(child_pid, SIGTERM);

    return ret == 0 ? child_ret : ret;
}
